File: alfaedge_finance/alfaedge_finance/page/bulk_payment_csv/bulk_payment_csv.py
# Copyright (c) 2026, alfaEdge
# Page: Bulk Payment CSV
#
# Whitelisted method that returns bulk-payment rows as JSON for the
# client-side editable table in bulk_payment_csv.js.
#
#   - Salary            -> Salary Slips under a selected Payroll Entry
#   - Expense Claims    -> unpaid claims, summed per employee
#   - Purchase Invoices -> overdue invoices, summed per supplier

import json
import re

import frappe
from frappe.utils import flt, formatdate

# Same roles as the page itself - the methods below expose salaries and bank account
# numbers, and whitelisted methods are otherwise callable by any logged-in user.
PAGE_ROLES = ("Accounts Manager", "System Manager")


@frappe.whitelist()
def get_payroll_entries(company):
	frappe.only_for(PAGE_ROLES)
	entries = frappe.get_all(
		"Payroll Entry",
		filters={"company": company, "docstatus": 1},
		fields=["name", "end_date"],
		order_by="end_date desc",
	)
	result = []
	for e in entries:
		label = frappe.utils.getdate(e.end_date).strftime("%B %Y") if e.end_date else e.name
		result.append({"name": e.name, "label": label})
	return result


@frappe.whitelist()
def get_rows(filters):
	frappe.only_for(PAGE_ROLES)
	if isinstance(filters, str):
		filters = json.loads(filters)
	filters = frappe._dict(filters or {})
	validate_filters(filters)

	debit_account_no = get_debit_account_no(filters.company)
	txn_date_str = formatdate(filters.transaction_date, "dd/MM/yyyy")

	if filters.txn_type == "Salary":
		return get_salary_rows(filters, debit_account_no, txn_date_str)
	elif filters.txn_type == "Expense Claims":
		return get_expense_claim_rows(filters, debit_account_no, txn_date_str)
	elif filters.txn_type == "Purchase Invoices":
		return get_purchase_invoice_rows(filters, debit_account_no, txn_date_str)
	return []


def validate_filters(filters):
	if not filters.company:
		frappe.throw("Company is mandatory")
	if not filters.transaction_date:
		frappe.throw("Transaction Date is mandatory")
	if not filters.txn_type:
		frappe.throw("Type is mandatory")
	if filters.txn_type == "Salary" and not filters.get("payroll_entry"):
		frappe.throw("Please select a Payroll Entry")


def get_debit_account_no(company):
	debit_account_no = frappe.db.get_value(
		"Bank Account",
		{"company": company, "is_default": 1, "is_company_account": 1},
		"bank_account_no",
	)
	if not debit_account_no:
		frappe.throw(
			f"No default company Bank Account (is_default & is_company_account both checked) "
			f"found for Company '{company}'"
		)
	return debit_account_no


def get_beneficiary_bank_account(party_type, party):
	fields = ["account_name", "bank_account_no", "branch_code"]
	bank_account = frappe.db.get_value(
		"Bank Account",
		{"party_type": party_type, "party": party, "is_default": 1},
		fields,
		as_dict=True,
	)
	if not bank_account:
		# fall back to any linked bank account if none is flagged as default
		bank_account = frappe.db.get_value(
			"Bank Account",
			{"party_type": party_type, "party": party},
			fields,
			as_dict=True,
		)
	return bank_account


def make_customer_ref_no(account_name, txn_date_str, bank_account_no):
	clean_name = re.sub(r"[^A-Za-z0-9]", "", account_name or "")
	prefix = (clean_name[:2] or "XX").upper()
	ddmm = "".join(txn_date_str.split("/")[:2])  # "DD" + "MM" from "DD/MM/YYYY"
	last3 = re.sub(r"[^0-9A-Za-z]", "", bank_account_no or "")[-3:]
	return f"{prefix}{ddmm}{last3}"


def make_nickname(beneficiary_name):
	return re.sub(r"[^A-Za-z0-9]", "", beneficiary_name or "").lower()


def build_row(debit_account_no, amount, bank_account, txn_date_str, party_label, reason=None):
	# Sums of floats drift (14320.229999999998) and this goes straight into the bank file.
	amount = flt(amount, 2)
	if not bank_account or reason:
		return {
			"matched": False,
			"reason": reason or "No Bank Account found",
			"party_label": party_label,
			"debit_account_no": debit_account_no,
			"transaction_amount": amount,
			"beneficiary_name": "",
			"beneficiary_account_no": "",
			"beneficiary_ifsc": "",
			"transaction_date": txn_date_str,
			"payment_mode": "N",
			"customer_ref_no": "",
			"beneficiary_nickname": "",
		}
	beneficiary_name = bank_account.account_name
	return {
		"matched": True,
		"party_label": party_label,
		"debit_account_no": debit_account_no,
		"transaction_amount": amount,
		"beneficiary_name": beneficiary_name,
		"beneficiary_account_no": bank_account.bank_account_no,
		"beneficiary_ifsc": bank_account.branch_code,
		"transaction_date": txn_date_str,
		"payment_mode": "N",
		"customer_ref_no": make_customer_ref_no(beneficiary_name, txn_date_str, bank_account.bank_account_no),
		"beneficiary_nickname": make_nickname(beneficiary_name),
	}


def get_salary_rows(filters, debit_account_no, txn_date_str):
	slips = frappe.get_all(
		"Salary Slip",
		filters={
			"payroll_entry": filters.payroll_entry,
			"company": filters.company,
			"docstatus": 1,
		},
		fields=["employee", "employee_name", "net_pay"],
	)

	rows = []
	for slip in slips:
		bank_account = get_beneficiary_bank_account("Employee", slip.employee)
		rows.append(
			build_row(debit_account_no, slip.net_pay, bank_account, txn_date_str,
				slip.employee_name or slip.employee)
		)
	return rows


def get_expense_claim_rows(filters, debit_account_no, txn_date_str):
	claims = frappe.get_all(
		"Expense Claim",
		filters={
			"company": filters.company,
			"docstatus": 1,
			"approval_status": "Approved",
		},
		fields=["employee", "employee_name", "grand_total", "total_advance_amount", "total_amount_reimbursed"],
	)

	totals = {}
	names = {}
	for c in claims:
		# Same outstanding as HRMS: sanctioned + taxes, less advances and reimbursements.
		outstanding = flt(c.grand_total) - flt(c.total_advance_amount) - flt(c.total_amount_reimbursed)
		if outstanding <= 0:
			continue
		totals[c.employee] = totals.get(c.employee, 0) + outstanding
		names[c.employee] = c.employee_name

	rows = []
	for employee, amount in totals.items():
		bank_account = get_beneficiary_bank_account("Employee", employee)
		rows.append(
			build_row(debit_account_no, amount, bank_account, txn_date_str, names.get(employee, employee))
		)
	return rows


def get_purchase_invoice_rows(filters, debit_account_no, txn_date_str):
	invoices = frappe.get_all(
		"Purchase Invoice",
		filters={
			"company": filters.company,
			"docstatus": 1,
			"outstanding_amount": [">", 0],
			# Due on the payment day counts too - otherwise it slips to the next run.
			"due_date": ["<=", filters.transaction_date],
			"on_hold": 0,
		},
		fields=["supplier", "supplier_name", "outstanding_amount", "currency"],
	)
	company_currency = frappe.get_cached_value("Company", filters.company, "default_currency")

	totals = {}
	names = {}
	foreign = {}
	for inv in invoices:
		names[inv.supplier] = inv.supplier_name
		if inv.currency != company_currency:
			# A USD invoice's outstanding is in USD - adding it to rupee totals would pay
			# $150 as Rs 150. Listed separately; these are paid by wire, not NEFT.
			key = (inv.supplier, inv.currency)
			foreign[key] = foreign.get(key, 0) + inv.outstanding_amount
			continue
		totals[inv.supplier] = totals.get(inv.supplier, 0) + inv.outstanding_amount

	rows = [
		build_row(
			debit_account_no,
			amount,
			None,
			txn_date_str,
			names.get(supplier, supplier),
			reason=f"Billed in {currency} - pay separately",
		)
		for (supplier, currency), amount in foreign.items()
	]
	advances = get_open_supplier_advances(filters.company, list(totals), filters.transaction_date)
	for supplier, amount in totals.items():
		label = names.get(supplier, supplier)
		advance = flt(advances.get(supplier), 2)
		if advance and amount - advance <= 0:
			rows.append(
				build_row(
					debit_account_no,
					amount,
					None,
					txn_date_str,
					label,
					reason=f"Covered by open advance of {advance:,.2f}",
				)
			)
			continue
		bank_account = get_beneficiary_bank_account("Supplier", supplier)
		row = build_row(debit_account_no, amount - advance, bank_account, txn_date_str, label)
		if advance:
			row.update(invoices_due=flt(amount, 2), advance_deducted=advance)
		rows.append(row)
	return rows


def get_open_supplier_advances(company, suppliers, up_to_date):
	"""Money already paid to each supplier and not yet set against an invoice - paying the
	full overdue total on top of it would overpay them. Payment Entries' unallocated amount,
	plus advance Journal Entry rows not linked to any invoice (both in company currency;
	only company-currency suppliers reach this)."""
	if not suppliers:
		return {}
	advances = {}
	for supplier, amount in frappe.db.sql(
		"""
		select party, sum(unallocated_amount)
		from `tabPayment Entry`
		where docstatus = 1 and company = %(company)s and payment_type = 'Pay'
			and party_type = 'Supplier' and party in %(suppliers)s
			and unallocated_amount > 0 and posting_date <= %(date)s
		group by party
		""",
		{"company": company, "suppliers": suppliers, "date": up_to_date},
	):
		advances[supplier] = advances.get(supplier, 0) + flt(amount)
	for supplier, amount in frappe.db.sql(
		"""
		select jea.party, sum(jea.debit - jea.credit)
		from `tabJournal Entry Account` jea
		join `tabJournal Entry` je on je.name = jea.parent
		where je.docstatus = 1 and je.company = %(company)s and je.posting_date <= %(date)s
			and jea.party_type = 'Supplier' and jea.party in %(suppliers)s
			and jea.is_advance = 'Yes' and ifnull(jea.reference_name, '') = ''
		group by jea.party
		""",
		{"company": company, "suppliers": suppliers, "date": up_to_date},
	):
		if flt(amount) > 0:
			advances[supplier] = advances.get(supplier, 0) + flt(amount)
	return advances


# The bank's template - column order and labels must match it exactly.
EXPORT_COLUMNS = [
	("Debit Account Number (Mandatory)", "debit_account_no"),
	("Transaction Amount (Mandatory)", "transaction_amount"),
	("Beneficiary Name (Mandatory)", "beneficiary_name"),
	("Beneficiary Account Number (Mandatory)", "beneficiary_account_no"),
	("Beneficiary IFSC Code (Mandatory)", "beneficiary_ifsc"),
	("Transaction Date (Mandatory)", "transaction_date"),
	("Payment Mode (Mandatory)", "payment_mode"),
	("Customer Reference Number (Mandatory)", "customer_ref_no"),
	("Beneficiary Nickname/Code (Mandatory)", "beneficiary_nickname"),
]


@frappe.whitelist()
def download_xlsx(rows, txn_type=None):
	"""Build the bank file on the server (openpyxl) from the rows as edited on the page -
	no spreadsheet library fetched from a CDN into a page showing salaries and account
	numbers. Every column except the amount is a text cell, so Excel or the bank's reader
	never turns a long account number into 9.2402E+14."""
	import io

	from openpyxl import Workbook

	frappe.only_for(PAGE_ROLES)
	rows = frappe.parse_json(rows) or []

	workbook = Workbook()
	sheet = workbook.active
	sheet.title = "Bulk Payment"
	sheet.append([label for label, _ in EXPORT_COLUMNS])
	for row in rows:
		values = []
		for _, fieldname in EXPORT_COLUMNS:
			value = row.get(fieldname)
			if fieldname == "transaction_amount":
				values.append(flt(value, 2))
			else:
				values.append("" if value is None else str(value))
		sheet.append(values)
	for column in sheet.iter_cols(min_row=2):
		for cell in column:
			if cell.data_type == "s":
				cell.number_format = "@"

	buffer = io.BytesIO()
	workbook.save(buffer)
	safe_type = re.sub(r"[^A-Za-z0-9]+", "_", txn_type or "Payments")
	frappe.response["filename"] = f"Bulk_Payment_{safe_type}_{frappe.utils.today()}.xlsx"
	frappe.response["filecontent"] = buffer.getvalue()
	frappe.response["type"] = "binary"

File: alfaedge_finance/alfaedge_finance/page/bulk_payment_csv/bulk_payment_csv.js
// Copyright (c) 2026, alfaEdge
// Page: Bulk Payment CSV

frappe.pages['bulk-payment-csv'].on_page_load = function (wrapper) {
	let page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Bulk Payment CSV',
		single_column: true,
	});

	new BulkPaymentCSV(page);
};

class BulkPaymentCSV {
	constructor(page) {
		this.page = page;
		this.rows = [];

		// column order/labels must match the bank's template EXACTLY
		this.columns = [
			{ label: "Debit Account Number (Mandatory)", fieldname: "debit_account_no" },
			{ label: "Transaction Amount (Mandatory)", fieldname: "transaction_amount" },
			{ label: "Beneficiary Name (Mandatory)", fieldname: "beneficiary_name" },
			{ label: "Beneficiary Account Number (Mandatory)", fieldname: "beneficiary_account_no" },
			{ label: "Beneficiary IFSC Code (Mandatory)", fieldname: "beneficiary_ifsc" },
			{ label: "Transaction Date (Mandatory)", fieldname: "transaction_date" },
			{ label: "Payment Mode (Mandatory)", fieldname: "payment_mode" },
			{ label: "Customer Reference Number (Mandatory)", fieldname: "customer_ref_no" },
			{ label: "Beneficiary Nickname/Code (Mandatory)", fieldname: "beneficiary_nickname" },
		];

		this.make_filters();
		this.make_table_area();
	}

	make_filters() {
		this.company_field = this.page.add_field({
			fieldname: 'company',
			label: 'Company',
			fieldtype: 'Link',
			options: 'Company',
			default: frappe.defaults.get_user_default('Company'),
			reqd: 1,
			change: () => this.load_payroll_entries(),
		});

		this.txn_type_field = this.page.add_field({
			fieldname: 'txn_type',
			label: 'Type',
			fieldtype: 'Select',
			options: ['Salary', 'Expense Claims', 'Purchase Invoices'],
			default: 'Expense Claims',
			reqd: 1,
			change: () => this.toggle_payroll_entry(),
		});

		this.transaction_date_field = this.page.add_field({
			fieldname: 'transaction_date',
			label: 'Transaction Date',
			fieldtype: 'Date',
			default: frappe.datetime.get_today(),
			reqd: 1,
		});

		this.page.set_primary_action('Generate', () => this.generate(), 'refresh');

		this.make_payroll_entry_selector();
		this.toggle_payroll_entry();
		this.load_payroll_entries();
	}

	// Built as a plain element (not page.add_field) so we don't depend on
	// internal Frappe control-wrapper property names for the HTML fieldtype.
	// Appended into page.page_form (the same row container the other three
	// fields live in) so it sits on the same line; falls back to page.body
	// if page_form isn't available in this Frappe version.
	// Clones the real "Type" Select field's markup so styling matches exactly
	// (guaranteed pixel-identical, rather than guessing Frappe's internal CSS
	// classes). Falls back to a manually built version if cloning fails.
	make_payroll_entry_selector() {
		let $clone = null;
		try {
			if (this.txn_type_field && this.txn_type_field.wrapper) {
				let $source = $(this.txn_type_field.wrapper).closest('[class*="col-"]');
				if ($source.length) {
					$clone = $source.clone();
					$clone.find('label, .control-label').remove();
					let $select = $clone.find('select');
					if ($select.length) {
						$select.attr('id', 'payroll-entry-select').empty().append('<option value="">Payroll Entry...</option>');
					} else {
						$clone = null;
					}
				}
			}
		} catch (e) {
			$clone = null;
		}

		this.payroll_entry_wrapper = $clone && $clone.length ? $clone : $(`
			<div class="col-sm-2">
				<div class="frappe-control" data-fieldtype="Select">
					<div class="control-input-wrapper">
						<div class="control-input">
							<select class="input-with-feedback form-control ellipsis" id="payroll-entry-select">
								<option value="">Payroll Entry...</option>
							</select>
						</div>
					</div>
				</div>
			</div>
		`);

		let target = (this.page.page_form && this.page.page_form.length) ? this.page.page_form : this.page.body;
		this.payroll_entry_wrapper.appendTo(target);

		this.$payroll_entry_select = this.payroll_entry_wrapper.find('#payroll-entry-select');
	}

	load_payroll_entries() {
		let company = this.company_field.get_value();
		if (!company) {
			this.$payroll_entry_select.html('<option value="">Select a Company first...</option>');
			return;
		}
		frappe.call({
			method: 'alfaedge_finance.alfaedge_finance.page.bulk_payment_csv.bulk_payment_csv.get_payroll_entries',
			args: { company },
			callback: (r) => {
				let entries = r.message || [];
				let options = ['<option value="">Payroll Entry...</option>']
					.concat(
						entries.map(
							(e) => `<option value="${frappe.utils.escape_html(e.name)}">${frappe.utils.escape_html(e.label)}</option>`
						)
					)
					.join('');
				this.$payroll_entry_select.html(options);
			},
		});
	}

	toggle_payroll_entry() {
		let show = this.txn_type_field.get_value() === 'Salary';
		this.payroll_entry_wrapper.toggle(show);
	}

	make_table_area() {
		this.table_wrapper = $(
			`<div class="bulk-payment-table-wrapper" style="margin-top: 20px; overflow-x: auto;"></div>`
		).appendTo(this.page.body);

		this.unmatched_wrapper = $(
			`<div class="bulk-payment-unmatched-wrapper" style="margin-top: 25px; overflow-x: auto;"></div>`
		).appendTo(this.page.body);
	}

	generate() {
		// Guard against the Date control returning a native JS Date object:
		// JSON-serializing a Date converts it to UTC, which rolls back to the
		// previous day for timezones ahead of UTC (like IST) — explicitly
		// normalize to a plain "YYYY-MM-DD" string to avoid that.
		let raw_date = this.transaction_date_field.get_value();
		let transaction_date = (raw_date instanceof Date) ? frappe.datetime.obj_to_str(raw_date) : raw_date;

		let filters = {
			company: this.company_field.get_value(),
			txn_type: this.txn_type_field.get_value(),
			transaction_date: transaction_date,
			payroll_entry: this.$payroll_entry_select.val(),
		};

		if (!filters.company || !filters.txn_type || !filters.transaction_date) {
			frappe.msgprint('Please fill Company, Type and Transaction Date');
			return;
		}
		if (filters.txn_type === 'Salary' && !filters.payroll_entry) {
			frappe.msgprint('Please select a Payroll Entry');
			return;
		}

		frappe.call({
			method: 'alfaedge_finance.alfaedge_finance.page.bulk_payment_csv.bulk_payment_csv.get_rows',
			args: { filters },
			freeze: true,
			freeze_message: 'Fetching records...',
			callback: (r) => {
				let all_rows = r.message || [];
				this.rows = all_rows.filter((row) => row.matched);
				this.unmatched_rows = all_rows.filter((row) => !row.matched);
				this.render_table();
			},
		});
	}

	render_table() {
		this.table_wrapper.empty();
		this.render_unmatched_table();

		if (!this.rows.length) {
			this.table_wrapper.html(
				'<div class="text-muted" style="padding: 20px;">No matched records found.</div>'
			);
			this.page.clear_secondary_action();
			return;
		}

		let header = this.columns.map((c) => `<th>${frappe.utils.escape_html(c.label)}</th>`).join('');
		let body = this.rows
			.map((row, i) => {
				let cells = this.columns
					.map((c) => {
						let val = row[c.fieldname] != null ? row[c.fieldname] : '';
						return `<td><input type="text" class="form-control input-sm bp-cell"
							data-row="${i}" data-field="${c.fieldname}"
							value="${frappe.utils.escape_html(String(val))}"></td>`;
					})
					.join('');
				return `<tr data-row="${i}">
					<td class="text-center"><button class="btn btn-xs btn-danger bp-remove-row" data-row="${i}" title="Remove row">&times;</button></td>
					${cells}
				</tr>`;
			})
			.join('');

		let $table = $(`
			<table class="table table-bordered table-sm" style="background:#fff; white-space: nowrap;">
				<thead><tr><th></th>${header}</tr></thead>
				<tbody>${body}</tbody>
			</table>
		`);
		this.table_wrapper.append($table);

		// keep this.rows in sync as the user edits cells
		$table.on('input change', '.bp-cell', (e) => {
			let $el = $(e.currentTarget);
			let i = $el.data('row');
			let field = $el.data('field');
			this.rows[i][field] = $el.val();
		});

		// remove a row and re-render (re-render keeps data-row indexes correct)
		$table.on('click', '.bp-remove-row', (e) => {
			let i = $(e.currentTarget).data('row');
			this.rows.splice(i, 1);
			this.render_table();
		});

		this.page.set_secondary_action('Download Excel', () => this.download_xlsx(), 'download');
	}

	render_advances_table() {
		let adjusted = (this.rows || []).filter((row) => row.advance_deducted);
		if (!adjusted.length) return;

		let fmt = (value) => frappe.utils.escape_html(format_number(value, null, 2));
		let rows_html = adjusted
			.map(
				(row) => `<tr>
					<td>${frappe.utils.escape_html(row.party_label || '')}</td>
					<td class="text-right">${fmt(row.invoices_due)}</td>
					<td class="text-right">${fmt(row.advance_deducted)}</td>
					<td class="text-right">${fmt(row.invoices_due - row.advance_deducted)}</td>
				</tr>`
			)
			.join('');
		this.unmatched_wrapper.append(`
			<div style="font-weight: 600; margin-bottom: 8px;">
				Open advances deducted (${adjusted.length}) — edit the amount above if an advance is meant for a later invoice
			</div>
			<table class="table table-bordered table-sm" style="background:#fafafa; margin-bottom: 20px;">
				<thead><tr><th>Supplier</th><th class="text-right">Invoices due</th><th class="text-right">Open advance</th><th class="text-right">Amount in file</th></tr></thead>
				<tbody>${rows_html}</tbody>
			</table>
		`);
	}

	render_unmatched_table() {
		this.unmatched_wrapper.empty();
		this.render_advances_table();

		if (!this.unmatched_rows || !this.unmatched_rows.length) {
			return;
		}

		this.unmatched_wrapper.append(
			`<div style="font-weight: 600; margin-bottom: 8px; color: #d13438;">
				Not included (${this.unmatched_rows.length}) — excluded from the CSV export
			</div>`
		);

		let rows_html = this.unmatched_rows
			.map(
				(row) => `<tr>
					<td>${frappe.utils.escape_html(row.party_label || '')}</td>
					<td>${frappe.utils.escape_html(String(row.transaction_amount != null ? row.transaction_amount : ''))}</td>
					<td>${frappe.utils.escape_html(row.reason || '')}</td>
				</tr>`
			)
			.join('');

		let $table = $(`
			<table class="table table-bordered table-sm" style="background:#fafafa;">
				<thead><tr><th>Employee / Supplier</th><th>Amount</th><th>Reason</th></tr></thead>
				<tbody>${rows_html}</tbody>
			</table>
		`);
		this.unmatched_wrapper.append($table);
	}

	download_xlsx() {
		if (!this.rows || !this.rows.length) {
			frappe.msgprint('No rows to export.');
			return;
		}
		// Built on the server from the rows as edited here - see download_xlsx in the .py.
		open_url_post(
			'/api/method/alfaedge_finance.alfaedge_finance.page.bulk_payment_csv.bulk_payment_csv.download_xlsx',
			{ rows: JSON.stringify(this.rows), txn_type: this.txn_type_field.get_value() }
		);
	}
}
